core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -5527,6 +5530,105 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
+        ZIO.withFiberRuntime[R, E, B1] { (fibRt, r) =>
+          ZIO
+            .asyncInterrupt[R, Nothing, Either[B, zio.Exit[E, A]]] { cb =>
+              //todo: optimize away when the timeout is <= 0 ?
+              val fib = ZIO.unsafe.makeChildFiber(
+                trace,
+                self,
+                fibRt,
+                r.runtimeFlags,
+                null
+              )(zio.Unsafe.unsafe)
+              import TimeoutTo._
+              val bypassState = new AtomicReference[BypassState[E, A]](BypassPossible)
+              val cancellable = scheduler
+                .schedule(
+                  () => {
+                    //race both with the fiber and the supervisor
+                    //fiber attempts to CAS state into BypassPendingResult while supervisor attempts to CAS into BypassDenied,
+                    //if we won the race or lost to the supervisor, we're now racing with the fiber.
+                    //otherwise, fiber existed early and the coordinator is still able to bypass cb
+                    if (bypassState.compareAndSet(BypassPossible, BypassDenied)) //won the race
+                      cb(ZIO.left(b()).ensuring(fib.interrupt *> fib.inheritAll))
+                    else if (bypassState.get /*Plain*/ eq BypassDenied) //supervisor won
+                      cb(ZIO.left(b()).ensuring(fib.interrupt *> fib.inheritAll))
+                  },
+                  duration
+                )(zio.Unsafe.unsafe)
+
+              if (
+                bypassState.get /*Plain*/ () eq BypassDenied
+              ) //no need to actually start the fiber (todo: remove it from the fibers scope?)
+                {
+                  fib.startSuspended()(zio.Unsafe.unsafe)
+                  Left(ZIO.unit)
+                } //todo: can we bypass here? it'd require the scheduler to change state into BypassPendingResult and adding a state so the scheduler does the right thing for 'late' timeout
+              else {
+                fib.addObserver { ex =>
+                  if (!bypassState.compareAndSet(BypassPossible, BypassPendingResult(ex))) {
+                    //lost the race, either to parent fiber or to the scheduler
+                    //in either case the state is BypassDenied and we're in race with the scheduler, and we know for sure this entire effect will be completed via cb.
+                    //since cb CAN be invoked multiple times, we simply delegate the race to cb
+                    // * notice that changing the scheduler to use bypass as well will require modifying this logic as state may be BypassPendingResult, violating the assumption behind this logic.
+                    cancellable.apply()
+                    cb {
+                      (fib.inheritAll.as(Right(ex)))
+                    }
+                  } //else: fiber won, parent now owns cb and can bypass it
+                }(zio.Unsafe.unsafe)
+                fib.start(self)
```

Comment:
@kyri-petrou @ghostdogpr ,
I suspect this creates a time-window in which the parent fiber doesn't respond to interrupts, 
this may happen when the `self` effect is cpu intensive and does not make any async calls.

is this a real concern?

Reply:
I was going through this PR earlier in the week and came to this point as well. Honestly, I don't really think adding timeouts to CPU-bound work makes sense, but I don't think we can't ignore the fact that it might indeed exist in the wild

Reply:
I actually used to have such a usecase where we executed 'untrusted' user code and wanted to make sure we can bail out in case to trouble (we did have some interrupt like mechanism to terminate the user code)